velox/exec/HashProbe.cpp

```diff
@@ -189,58 +225,276 @@ void HashProbe::initializeFilter(
-    // Anti join with null keys on the build side always returns nothing.
+    // Null-aware anti join with null keys on the build side without a filter
+    // always returns nothing.
     VELOX_CHECK(isNullAwareAntiJoin(joinType_));
-    finished_ = true;
-  } else {
-    table_ = hashBuildResult->table;
-    if (table_->numDistinct() == 0) {
-      // Build side is empty. Inner, right and semi joins return nothing in this
-      // case, hence, we can terminate the pipeline early.
-      if (isInnerJoin(joinType_) || isLeftSemiJoin(joinType_) ||
-          isRightJoin(joinType_) || isRightSemiJoin(joinType_)) {
-        finished_ = true;
+    // 'antiJoinHasNullKeys' flag shall only be set on the first built 'table_'
+    // and we don't expect any spilled probe data at that point.
+    VELOX_CHECK(spillPartitionSet_.empty());
+    spillPartitionSet_.clear();
+    noMoreInput();
+    return;
+  }
+
+  table_ = std::move(hashBuildResult->table);
+  VELOX_CHECK_NOT_NULL(table_);
+
+  maybeSetupSpillInput(
+      hashBuildResult->restoredPartitionId, hashBuildResult->spillPartitionIds);
+
+  if (table_->numDistinct() == 0) {
+    // If the build table is empty, then inner, right and semi joins return
+    // nothing. Given that, we can call 'noMoreInput()' to skip processing the
+    // probe inputs. But if there is spilling triggered at the build side while
+    // building the table, then we still need to process probe inputs to spill
+    // the probe rows if the corresponding partitions have been spilled at the
+    // build side.
+    if (skipProbeOnEmptyBuild()) {
+      if (!needSpillInput()) {
+        noMoreInput();
       }
-    } else if (
-        (isInnerJoin(joinType_) || isLeftSemiJoin(joinType_) ||
-         isRightSemiJoin(joinType_)) &&
-        table_->hashMode() != BaseHashTable::HashMode::kHash) {
-      // Find out whether there are any upstream operators that can accept
-      // dynamic filters on all or a subset of the join keys. Create dynamic
-      // filters to push down.
-      const auto& buildHashers = table_->hashers();
-      auto channels = operatorCtx_->driverCtx()->driver->canPushdownFilters(
-          this, keyChannels_);
-      for (auto i = 0; i < keyChannels_.size(); i++) {
-        if (channels.find(keyChannels_[i]) != channels.end()) {
-          if (auto filter = buildHashers[i]->getFilter(false)) {
-            dynamicFilters_.emplace(keyChannels_[i], std::move(filter));
-          }
+    }
+  } else if (
+      (isInnerJoin(joinType_) || isLeftSemiJoin(joinType_) ||
+       isRightSemiJoin(joinType_)) &&
+      table_->hashMode() != BaseHashTable::HashMode::kHash && !isSpillInput() &&
+      !hasMoreSpillData()) {
+    // Find out whether there are any upstream operators that can accept
+    // dynamic filters on all or a subset of the join keys. Create dynamic
+    // filters to push down.
+    //
+    // NOTE: this optimization is not applied in the following cases: (1) if the
```

Comment:
Curious if we have test cases to ensure dynamic filtering is disabled in case of spilling.

Reply:
Yes, we check dynamic filtering is not set in existing HashJoinTestCase with spilling enabled.